### PR TITLE
credentials: improved experience for profile creation

### PR DESCRIPTION
**Pull request by breedloj**
## Problem
Users routinely complain of difficulty in setting up credentials profile in the Toolkit. The process requires some knowledge of how to create access keys and a guided flow for providing them to the toolkit.

## Solution
This change contains several UX improvements to walk users through the process of setting up a credentials profile and providing them quick access to documentation resources. It also fixes an issue where profiles would not be automatically connected once created.

### Credentials creation wizard changes
Added a help link to the credentials creation prompt, linking out to the [credentials setup documentation](https://docs.aws.amazon.com/toolkit-for-vscode/latest/userguide/setup-credentials.html).
image

Added a help icon above credentials setup steps that also links out to the credentials setup documentation. Notably, clicking this does **not** interrupt the wizard flow and users can come back and resume where they left off.
createProfile

Added input validation for empty values across profile name, access key, and secret key. Previously it was possibly to quickly enter through these steps with no input.
validation

Added a prompt when a user successfully creates a credentials profile. Previously there was no obvious visual cue that they'd accomplished something other than the extension features becoming available.
Screen Shot 2021-07-11 at 12 47 49 PM

Additionally, this change fixes a bug where a newly created profile would not be automatically connected to and would require users to manually select the profile.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1829,6 +1829,37 @@
                     "separator": "\\"
                 }
             }
+        ],
+        "walkthroughs": [
+            {
+                "id": "getStarted",
+                "title": "%AWS.walkthrough.gettingStarted.title%",
+                "description": "%AWS.walkthrough.gettingStarted.description%",
+                "when": "!config.aws.profile",
```
Comment (bryceitoc9) [inline package.json:1838]: Does this mean that this will always show up if a user has no AWS profile? Our functionality doesn't depend on AWS credentials (e.g. a user who only uses the toolkit for SAM debugging, and deploys via CI), so this can get annoying if that's the case.

File: package.json [hunk 2]
```diff
@@ -1829,6 +1829,37 @@
...
+                "steps": [
+                    {
+                        "id": "accessKeys",
+                        "title": "%AWS.walkthrough.gettingStarted.accessKeys.title%",
+                        "description": "%AWS.walkthrough.gettingStarted.accessKeys.description%",
+                        "media": {
+                            "image": "resources/walkthrough/iamCredentials.png",
+                            "altText": "Empty image"
```
Comment (bryceitoc9) [inline package.json:1846]: We should probably have something more descriptive here, I'm assuming this would be used by screen readers, etc. Should this also be localized/can it be?

File: src/shared/credentials/defaultCredentialSelectionDataProvider.ts [hunk 3]
```diff
@@ -59,62 +69,91 @@ export class DefaultCredentialSelectionDataProvider implements CredentialSelecti
...
+            return localize('AWS.credentials.error.emptyProfileName', 'Profile name must not be empty')
+        }
+
         const duplicate = this.existingProfileNames.find(k => k === name)
 
         return duplicate ? 'Name not unique' : undefined
     }
 
     public async validateAccessKey(accessKey: string): Promise<string | undefined> {
         // TODO: is there a regex pattern we could use?
```
Comment (bryceitoc9) [inline src/shared/credentials/defaultCredentialSelectionDataProvider.ts:145]: Not your comment, but you can use the git-secrets regex (see the `--register-aws` flag: https://github.com/awslabs/git-secrets#operation-modes . Unsure if there's a secret key flag, but there definitely is for access.

File: src/shared/defaultAwsContextCommands.ts [hunk 4]
```diff
@@ -150,7 +145,20 @@ export class DefaultAWSContextCommands {
...
 
-                return state.profileName
+                const sharedProviderId: CredentialsId = {
+                    credentialSource: SharedCredentialsProvider.getProviderType(),
+                    credentialTypeId: state.profileName,
+                }
+
+                window.showInformationMessage(
+                    localize(
+                        'AWS.message.prompt.credentials.definition.done',
```
Comment (bryceitoc9) [inline src/shared/defaultAwsContextCommands.ts:155]: Doesn't look like this is in package.nls.json

File: src/shared/defaultAwsContextCommands.ts [hunk 5]
```diff
@@ -190,43 +197,38 @@ export class DefaultAWSContextCommands {
...
+                getIdeProperties().company
+            ),
+            localizedText.yes,
+            localizedText.no,
+            localizedText.help
+        )
+
+        if (userResponse === localizedText.yes) {
+            return true
+        } else if (userResponse === localizedText.help) {
```
Comment (bryceitoc9) [inline src/shared/defaultAwsContextCommands.ts:227]: Does this close the notification? Doesn't do a lot of good if the user clicks this and can't find their way back.

File: package.json (hunk 2 above)
Comment (JadenSimon) [reply package.json:1846]: Yeah 'altText' refers to accessibility in this case, and it should be localized.

File: src/shared/defaultAwsContextCommands.ts (hunk 4 above)
Comment (JadenSimon) [reply src/shared/defaultAwsContextCommands.ts:155]: Does not need to be in `package.nls.json`
The real localization will extract it out.

File: src/shared/defaultAwsContextCommands.ts (hunk 5 above)
Comment (JadenSimon) [reply src/shared/defaultAwsContextCommands.ts:227]: Does not close it. Flow controller was modified.

File: package.json (hunk 1 above)
Comment (JadenSimon) [reply package.json:1838]: The walkthrough is not super intrusive. It just shows up on the start screen (no prompts/notifications/etc.)

File: package.json (hunk 2 above)
Comment (bryceitoc9) [reply package.json:1846]: I wouldn't be surprised if VS Code lets you pull a remote image (in which this is the placeholder). But yeah, anything is better than this.

Comment (breedloj) [reply package.json]: You're both correct. And interestingly they do allow remote images here, though in this case I'm going to embed them. Added some more meaningful text sourced through the NLS file.

File: src/shared/credentials/defaultCredentialSelectionDataProvider.ts (hunk 3 above)
Comment (breedloj) [reply src/shared/credentials/defaultCredentialSelectionDataProvider.ts:145]: Interesting, good find. I am hesitant to use this however at least in the current blocking validation we do. Even within the docs for git-secrets it's called out that...

`While the patterns registered by this command should catch most instances of AWS credentials, these patterns are not guaranteed to catch them all.`

Unless we can get authoritative regexes from the IAM team we probably wouldn't want to do more than provide guidance ("Hey, it looks like your access key may be invalid. Might want to check on that.") vs. inadvertently blocking something that is actually correct (though even the aforementioned guidance could lead to confusion).

File: src/shared/defaultAwsContextCommands.ts (hunk 5 above)
Comment (breedloj) [reply src/shared/defaultAwsContextCommands.ts:227]: Updated this logic to reprompt if they hit help so for all intents and purposes they return to what appears to be the original prompt. Kind of a goofy way to do it but VSCode does not allow onClick event handling that does not also close the prompt.

File: package.json (hunk 1 above)
Comment (breedloj) [reply package.json:1838]: Hmm, interesting. I don't have a sense for how common those cases are, or whether those users would have a desire to graduate to the 'full' toolkit experience to which this walkthrough would be helpful. But to Jaden's point the walkthrough is not prominent and can be hidden. This when clause just controls the conditions in which the walkthrough would originally be surfaced through the welcome screen. I am open to suggestions on a better hook but I also don't want the conditions to be too narrow, i.e.: only on the first 'run' of the extension via a context setting, since a user may go months without using the Toolkit, end up with now invalid credentials, and has since forgotten the setup procedure.

File: src/shared/defaultAwsContextCommands.ts (hunk 4 above)
Comment (breedloj) [reply src/shared/defaultAwsContextCommands.ts:155]: The pattern I have been following since an earlier PR in which I received feedback is to only add explicit NLS file entries when adding references in `package.json`.

File: src/shared/credentials/defaultCredentialSelectionDataProvider.ts (hunk 3 above)
Comment (bryceitoc9) [reply src/shared/credentials/defaultCredentialSelectionDataProvider.ts:145]: Perhaps shorten this to `[A-Z0-9]{20}` and just cut the prefixes?

Comment (breedloj) [reply src/shared/credentials/defaultCredentialSelectionDataProvider.ts:145]: Even this I'd be hesitant to add since as far as I can find this isn't a documented length constraint (the API states a [minimum of 16 and max of 128 characters](https://docs.aws.amazon.com/IAM/latest/APIReference/API_AccessKey.html)) and such a constraint risks being forward incompatible.

**Review by bryceitoc9 — APPROVED** [review]
Only other consideration I'd make would be to add this to the extension readme in addition to or instead of the VS Code walkthrough, since it's already a mechanism we're using, and it applies to all IDEs.

**Review by JadenSimon — APPROVED** [review]
One other thing:
For first-time installs, the walkthrough will show up alongside all the other notifications (telemetry, show quick start, etc.)
This could be a bit overwhelming, but I'm not sure what the best way is to combine them

**Comment by breedloj** [comment]
Removed the walkthrough..for now. 😃

**Review by JadenSimon — APPROVED** [review]

**Comment by JadenSimon** [comment]
Test failure unrelated

**Review by gsweetwood — APPROVED** [review]
